app/assets/javascripts/discourse/app/mixins/uppy-upload.js

```diff
@@ -320,21 +322,30 @@ export default Mixin.create(UppyS3Multipart, {
   _bindFileInputChange() {
     this.fileInputEventListener = bindFileInputChangeListener(
       this.fileInputEl,
-      (file) => {
-        try {
-          this._uppyInstance.addFile({
-            source: `${this.id} file input`,
+      this._addFiles
+    );
+  },
+
+  @bind
+  _addFiles(files, opts = {}) {
```

Comment:
How is the first parameter (files) passed in by `this.appEvents.on` / `this.appEvents.off`? Is it because we're `@bind`-ing it to the caller's scope? If that's the case `off` doesn't have files and is it fine?

Reply:
`appEvents` is a service that is automatically injected into mixins/components. When you `trigger` an appEvent you pass through the params you want

https://github.com/discourse/discourse/blob/9366a5862cea9a66a0ff11f30544b1f343d4c627/app/assets/javascripts/discourse/tests/integration/components/emoji-uploader-test.js#L58-L60

`@bind` is used to make sure the scope inside `_addFiles` still refers to the component/mixin (you can also do `    this.appEvents.on(``upload-mixin:${this.id}:add-files``, this._addFiles.bind(this));`)